server/src/main/java/org/opensearch/search/optimization/filterrewrite/OptimizationContext.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.optimization.filterrewrite;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.index.DocValues;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.NumericDocValues;
+import org.apache.lucene.index.PointValues;
+import org.opensearch.index.mapper.DocCountFieldMapper;
+import org.opensearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.function.BiConsumer;
+
+import static org.opensearch.search.optimization.filterrewrite.Helper.loggerName;
+import static org.apache.lucene.search.DocIdSetIterator.NO_MORE_DOCS;
+
+/**
+ * Context object for doing the filter rewrite optimization in ranges type aggregation
+ * <p>
+ * This holds the common business logic and delegate aggregator-specific logic to {@link AggregatorBridge}
+ *
+ * @opensearch.internal
+ */
+public final class OptimizationContext {
+
+    private static final Logger logger = LogManager.getLogger(loggerName);
+
+    private boolean canOptimize = false;
+    private boolean preparedAtShardLevel = false;
+
+    final AggregatorBridge aggregatorBridge;
+    int maxAggRewriteFilters;
+    String shardId;
+
+    private Ranges ranges;
+    private Ranges rangesFromSegment;
+
+    // debug info related fields
+    private int leaf;
+    private int inner;
+    private int segments;
+    private int optimizedSegments;
+
+    public OptimizationContext(AggregatorBridge aggregatorBridge) {
+        this.aggregatorBridge = aggregatorBridge;
+    }
+
+    public boolean canOptimize(final Object parent, final int subAggLength, SearchContext context) {
+        if (context.maxAggRewriteFilters() == 0) return false;
+
+        if (parent != null || subAggLength != 0) return false;
+
+        this.canOptimize = aggregatorBridge.canOptimize();
+        if (canOptimize) {
+            aggregatorBridge.setOptimizationContext(this);
+            this.maxAggRewriteFilters = context.maxAggRewriteFilters();
+            this.shardId = context.indexShard().shardId().toString();
+        }
+        logger.debug("Fast filter rewriteable: {} for shard {}", canOptimize, shardId);
+        return canOptimize;
+    }
+
+    public void prepare() throws IOException {
```

Comment:
prepare is invoked from each Aggregator right after calling `canOptimize`, you can make this private and invoke directly from `canOptimize`.

Reply:
Make sense